File: fibhowmany.hs

range n = [0..n-1]

all_d n =
    if n == 0
        then [[]]
    else
        let
            before = all_d (n - 1)
        in
            [[x] ++ y | x <- range 10 , y <- before]

to_test = all_d 5

tablify [] = [0, 0, 0, 0, 0,  0, 0, 0, 0, 0]
tablify (x : xs) = 
    let (a, y : b) = splitAt x (tablify xs)
        in a ++ [y + 1] ++ b

check_fib nums = let
        table = tablify nums 
    in
        flip any [(a,b) | a <- range 10, b <- range 10] $ \(a, b) ->
            let
                follow table' a b =
                    if (sum table') == 0
                       then True
                       else if (table' !! a) <= 0
                                then False
                                else let
                                    (u, y : v) = splitAt a table'
                                    table'' = u ++ [y-1] ++ v
                                  in
                                    follow table'' b ((a+b) `mod` 10)
            in
                follow table a b
        
total = sum [
    (
        if check_fib s
           then 1
           else 0
       )
           
       | s <- to_test
    ]

main = print total

File: fibhowmany.py

def all_d(n):
    if n == 0:
        return [[]]
    else:
        before = all_d(n - 1)
        return [[x] + y for x in range(10) for y in before]

to_test = all_d(5)

def check_fib(nums):
    for a in range(10):
        for b in range(10):
            table = [0] * 10
            for i in nums:
                table[i] += 1
            
            def follow(a, b):
                if sum(table) == 0:
                    return True
                else:
                    if table[a] <= 0:
                        return False
                    else:
                        table[a] -= 1
                        return follow(b, (a+b) % 10)
            
            if follow(a, b):
                return True
    
    return False

total = 0
for i, s in zip(range(len(to_test)), to_test):
    if i % 100 == 0:
        print(i)
    if check_fib(s):
        total += 1
        
